sssd_test_framework/utils/sbus.py

```diff
@@ -0,0 +1,672 @@
+
+        result.text = result.text.removeprefix(res.group(0))
+        key = cls.get_python_value(result)
+        value = cls.get_python_value(result)
+
+        res = re.match(r"^ *\) *", result.text)
+        if res is None:
+            raise TypeError("Malformed dictionary entry")
+
+        result.text = result.text.removeprefix(res.group(0)).strip()
+        return key, value
+
+    @classmethod
+    def parse(cls, result: DBUSResult) -> DBUSTypeDict:
+        res = re.match(r"^ *array +\[ *(dict entry *\(.+\)) *\] *", result.text)
+        if res is None:
+            raise TypeError("Not a dictionary")
+
+        res2 = DBUSResult(res.group(1).strip())
+        dict = {}
+        while len(res2.text) > 0:
+            key, value = cls._parse_dict_entry(res2)
+            dict[key] = value.value if isinstance(value, DBUSTypeVariant) else value
+
+        result.text = result.text[res.span()[1] :].strip()
+        return DBUSTypeDict(dict)
+
+
+class DBUSObject(MultihostUtility[MultihostHost]):
+    """
+    D-Bus access.
+
+    Objects of this class represent the root or child D-Bus nodes.
+    """
+
+    class ProxyObject(ABC):
+        """
+        Abstract class representing a D-Bus object (method or property).
+
+        Corresponding concrete classes are provided for their instantiation.
+        """
+
+        def __init__(
+            self,
+            host: MultihostHost,
+            dest: str,
+            objpath: str,
+            bus: str | DBUSBus,
+            /,
+            interface: str | None = None,
+            child: xml.Element | None = None,
+        ):
+            """
+            Constructor to be called by the subclasses either explicitely or by inheritance.
+            """
+            match bus:
+                case DBUSBus.SYSTEM:
+                    _bus_param = "--system"
+                case DBUSBus.SESSION:
+                    _bus_param = "--session"
+                case _:
+                    _bus_param = "--bus=" + bus
+
+            self.host = host
+            self.dest = dest
+            self.objpath = objpath
+            self.bus = bus
+            self.interface = objpath[1:].replace("/", ".") if interface is None else interface
+            self.child = child
+            self.cmd = f"dbus-send {_bus_param} --print-reply --dest={dest} {self.objpath}"
+
+        def _run(self, command: str, *args):
+            """
+            Run dbus-send over ssh on the remote host.
+            """
+            cmd = self.cmd + f" {command}"
+            for param in args:
+                cmd += " " + DBUSType.get_dbus_param(param)
+
+            rslt = self.host.ssh.run(cmd)
+            if rslt.stdout_lines[0].startswith("method return"):
+                lines = rslt.stdout_lines[1:]
+            else:
+                lines = rslt.stdout_lines
+
+            return SSHProcessResult(rslt.rc, lines, rslt.stderr_lines)
+
+    class ProxyProperty(ProxyObject):
```

Comment:
I like what you did with properties, but it is also little problematic. D-Bus allows to have the same name for property and method or signal. Nothing prevents that.

Let's drop this code for now (keep a copy though), focus on minimizing and improving the rest of this internals. After that we will see if and how can we bring it back. If it won't fit, we will just add it by calling org.freedesktop.dbus.property interface similar way you add ifp now.